src/wallet/wallet.cpp

```diff
@@ -3335,3 +3335,330 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
+                    }
+                    dests_to_delete.push_back(addr_pair.first);
+                    continue;
+                }
+            }
+            // Not ours, not in watchonly wallet, and not in solvable
+            error = _("Error: Address book data in wallet cannot be identified to belong to migrated wallets");
+            return false;
+        }
+    }
+    // Remove the things to delete
+    if (dests_to_delete.size() > 0) {
+        for (const auto& dest : dests_to_delete) {
+            if (!DelAddressBook(dest)) {
+                error = _("Error: Unable to remove watchonly address book data");
+                return false;
+            }
+        }
+    }
+
+    WalletLogPrintf("Wallet migration complete.\n");
+
+    return true;
+}
+
+bool MigrateLegacyToDescriptor(CWallet& wallet, WalletContext& context, bilingual_str& error)
+{
+    LOCK(wallet.cs_wallet);
+
+    // First change to using SQLite
+
+    // Remove this wallet from the list of wallets while we mess with the database
+    // This is a workaround to prevent MaybeCompactWalletDB from accessing m_database while we are deleting and changing it
+    std::shared_ptr<CWallet> this_wallet{nullptr};
+    {
+        LOCK(context.wallets_mutex);
+        std::vector<std::shared_ptr<CWallet>>::iterator i = std::find_if(context.wallets.begin(), context.wallets.end(), [&wallet](std::shared_ptr<CWallet> w) {
+                return w.get() == &wallet;
+            });
+        assert(i != context.wallets.end());
+        this_wallet = *i;
+        context.wallets.erase(i);
+    }
+    // Now do the database stuff
+    if (!wallet.MigrateToSQLite(error)) return false;
+    // Put the wallet back in vpwallets because we are now done with messing with m_database
+    {
+        LOCK(context.wallets_mutex);
+        context.wallets.push_back(this_wallet);
+    }
+
+    // Then get all of the descriptors from the legacy wallet
+    std::optional<MigrationData> data = wallet.GetDescriptorsForLegacy(error);
+    if (data == std::nullopt) return false;
+
+    // Create the watchonly and solvable wallets if necessary
+    if (data->watch_descs.size() > 0 || data->solvable_descs.size() > 0) {
+        DatabaseOptions options;
+        options.require_existing = false;
+        options.require_create = true;
+
+        // Make the wallets
+        options.create_flags = WALLET_FLAG_DISABLE_PRIVATE_KEYS | WALLET_FLAG_BLANK_WALLET | WALLET_FLAG_DESCRIPTORS;
+        if (wallet.IsWalletFlagSet(WALLET_FLAG_AVOID_REUSE)) {
+            options.create_flags |= WALLET_FLAG_AVOID_REUSE;
+        }
+        if (wallet.IsWalletFlagSet(WALLET_FLAG_KEY_ORIGIN_METADATA)) {
+            options.create_flags |= WALLET_FLAG_KEY_ORIGIN_METADATA;
+        }
+        if (data->watch_descs.size() > 0) {
+            wallet.WalletLogPrintf("Making a new watchonly wallet containing the watched scripts\n");
+
+            DatabaseStatus status;
+            std::vector<bilingual_str> warnings;
```

Comment:
well, now you are ignoring those?

Reply:
I don't think that those warnings matter. There aren't any warnings for newly created wallets. These were being ignored originally anyways.